relay/writer/file.go

```diff
@@ -0,0 +1,381 @@
+func (w *FileWriter) handleRotateEvent(ev *replication.BinlogEvent) (*Result, error) {
+	rotateEv, ok := ev.Event.(*replication.RotateEvent)
+	if !ok {
+		return nil, errors.NotValidf("except RotateEvent, but got %+v", ev.Header)
+	}
+
+	// update binlog filename if needed
+	var currFile = w.filename.Get()
+	nextFile := string(rotateEv.NextLogName)
+	if nextFile > currFile {
+		// record the next filename, but not create it.
+		// even it's a fake RotateEvent, we still need to record it,
+		// because if we do not specify the filename when creating the writer (like Auto-Position),
+		// we can only receive a fake RotateEvent before the FormatDescriptionEvent.
+		w.filename.Set(nextFile)
+	}
+
+	// write the RotateEvent if not fake
+	var ignore bool
+	if ev.Header.Timestamp == 0 || ev.Header.LogPos == 0 {
+		// skip fake rotate event
+		ignore = true
+	} else if w.out == nil {
+		// if not open a binlog file yet, then non-fake RotateEvent can't be handled
+		return nil, errors.Errorf("non-fake RotateEvent %+v received, but no binlog file opened", ev.Header)
+	} else {
+		err := w.out.Write(ev.RawData)
+		if err != nil {
+			return nil, errors.Annotatef(err, "write RotateEvent %+v for %s", ev.Header, filepath.Join(w.cfg.RelayDir, currFile))
+		}
+	}
+
+	// update offset if the event is wrote
+	if !ignore {
+		w.offset.Add(int64(len(ev.RawData)))
+	}
+
+	return &Result{
+		Ignore: ignore,
+	}, nil
+}
+
+// handle non-special event:
+//   1. handle a potential hole if exists
+//   2. handle any duplicate events if exist
+//   3. write the non-duplicate event and update the offset
+func (w *FileWriter) handleEventDefault(ev *replication.BinlogEvent) (*Result, error) {
+	// handle a potential hole
+	err := w.handleFileHoleExist(ev)
+	if err != nil {
+		return nil, errors.Annotatef(err, "handle a potential hole in %s before %+v",
+			w.filename.Get(), ev.Header)
+	}
+
+	// handle any duplicate events if exist
+	result, err := w.handleDuplicateEventsExist(ev)
+	if err != nil {
+		return nil, errors.Annotatef(err, "handle a potential duplicate event %+v in %s",
+			ev.Header, w.filename.Get())
+	}
+	if result.Ignore {
+		// duplicate, and can ignore it. now, we assume duplicate events can all be ignored
+		return result, nil
+	}
+
+	// write the non-duplicate event and update the offset
+	err = w.out.Write(ev.RawData)
+	if err == nil {
+		w.offset.Add(int64(len(ev.RawData)))
+	}
+	return &Result{
+		Ignore: false,
+	}, errors.Annotatef(err, "write event %+v", ev.Header)
+}
+
+// handleFileHoleExist tries to handle a potential hole after this event wrote.
+// A hole exists often because some binlog events not sent by the master.
+// NOTE: handle cases when file size > 4GB
+func (w *FileWriter) handleFileHoleExist(ev *replication.BinlogEvent) error {
+	// 1. detect whether a hole exists
+	evStartPos := int64(ev.Header.LogPos - ev.Header.EventSize)
+	fileOffset := w.offset.Get()
+	holeSize := evStartPos - fileOffset
+	if holeSize <= 0 {
```

Comment:
does `< 0` implie error :thinking: ?

Reply:
`< 0 ` implies duplicate events may exist, that should be handled in `handleDuplicateEventsExist`, in this function (`handleFileHoleExist`) we only handle a potential hole.

Reply:
please add a comment

Reply:
there is a comment below this line.